File: src/bot/rules.py
"""Pure business rules for the trade confirmation bot."""

import re

from .models import CommentData, ConfirmationContext, ValidationResult

FLAIR_PATTERN = re.compile(r"Trades: (\d+)")
FLAIR_TEMPLATE_PATTERN = re.compile(r"Trades: ((\d+)-(\d+))")


def is_confirming_trade(comment_body: str) -> bool:
    """Check if comment is confirming a trade."""
    return "confirmed" in comment_body.lower()


def build_confirmation_key(parent_comment_id: str | None, confirmer: str) -> str:
    """Build the idempotency key used for paired flair increments."""
    return f"{parent_comment_id}:{confirmer}".lower()


def parse_trade_count(flair_text: str | None) -> int | None:
    """Extract tracked trade count from flair text."""
    if not flair_text:
        return 0

    match = FLAIR_PATTERN.search(flair_text)
    return int(match.group(1)) if match else None


def format_flair_from_template(flair_template: str, count: int) -> str:
    """Replace the tracked trade range in a flair template with the exact count."""
    match = FLAIR_TEMPLATE_PATTERN.search(flair_template)
    if not match:
        return flair_template

    start, end = match.span(1)
    return flair_template[:start] + str(count) + flair_template[end:]


def should_include_comment(
    *,
    submission_id: str,
    current_submission_id: str,
    is_root: bool,
    body_lower: str,
) -> bool:
    """Decide whether a polled comment should be included for processing.

    Only called for comments that are on an active submission, not banned,
    and from a processable redditor.
    """
    if is_root:
        return submission_id != current_submission_id

    return "confirmed" in body_lower or "approved" in body_lower


def is_possible_watermark_gap(
    *,
    had_initial_watermark: bool,
    found_seen: bool,
    listing_exhausted: bool,
    scanned_count: int,
    gap_threshold: int,
) -> bool:
    """Detect when scanning reached the listing limit without finding known IDs."""
    return (
        had_initial_watermark
        and not found_seen
        and listing_exhausted
        and scanned_count >= gap_threshold
    )


def evaluate_confirmation(
    comment: CommentData,
    context: ConfirmationContext,
) -> ValidationResult:
    """Pure validation of a confirmation comment given pre-fetched context."""
    if comment.is_root:
        return ValidationResult(valid=False)

    if not context.parent_exists or context.parent_is_banned:
        return ValidationResult(valid=False)

    if not context.parent_is_processable:
        return ValidationResult(valid=False)

    if context.parent_author_name == comment.author_name:
        return ValidationResult(valid=False)

    comment_body = comment.body.lower()

    # Mod approval path: comment is a reply to a confirmation (non-root parent).
    if not context.parent_is_root:
        if "approved" in comment_body and context.is_moderator:
            if context.grandparent_exists and context.grandparent_is_root:
                return ValidationResult(
                    valid=True,
                    is_mod_approval=True,
                    parent_author=context.grandparent_author_name,
                    confirmer=context.parent_author_name,
                    parent_comment_id=context.grandparent_id,
                    reply_to_comment_id=context.parent_id,
                )
        return ValidationResult(valid=False)

    if not is_confirming_trade(comment_body):
        return ValidationResult(valid=False)

    if context.parent_is_saved:
        return ValidationResult(
            valid=False,
            reason="already_confirmed",
            parent_author=context.parent_author_name,
            parent_comment_id=context.parent_id,
        )

    username_lower = comment.author_name.lower()
    if (
        username_lower not in context.parent_body_lower
        and username_lower not in context.parent_body_html_lower
    ):
        return ValidationResult(
            valid=False,
            reason="cant_confirm_username",
            parent_author=context.parent_author_name,
            parent_body_lower=context.parent_body_lower,
            parent_body_html_lower=context.parent_body_html_lower,
            confirmer=username_lower,
        )

    return ValidationResult(
        valid=True,
        parent_author=context.parent_author_name,
        confirmer=comment.author_name,
        parent_comment_id=context.parent_id,
        reply_to_comment_id=comment.id,
    )


def find_flair_template(
    templates: dict[tuple[int, int], dict],
    trade_count: int,
    is_moderator: bool,
) -> dict | None:
    """Find the flair template matching the given trade count and mod status."""
    for (min_trades, max_trades), template in templates.items():
        if min_trades <= trade_count <= max_trades:
            if template["mod_only"] == is_moderator:
                return template
    return None

File: internal/rules/rules.go
package rules

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/mikeacjones/reddit-trade-confirmation-bot/internal/models"
)

var (
	flairPattern         = regexp.MustCompile(`Trades: (\d+)`)
	flairTemplatePattern = regexp.MustCompile(`Trades: ((\d+)-(\d+))`)
)

// IsConfirmingTrade checks if a comment body is confirming a trade.
func IsConfirmingTrade(commentBody string) bool {
	return strings.Contains(strings.ToLower(commentBody), "confirmed")
}

// BuildConfirmationKey builds the idempotency key used for paired flair increments.
func BuildConfirmationKey(parentCommentID, confirmer string) string {
	return strings.ToLower(parentCommentID + ":" + confirmer)
}

// ParseTradeCount extracts tracked trade count from flair text.
// Returns (0, true) for empty flair, (n, true) for tracked, (0, false) for untracked custom flair.
func ParseTradeCount(flairText *string) (count int, tracked bool) {
	if flairText == nil || *flairText == "" {
		return 0, true
	}
	match := flairPattern.FindStringSubmatch(*flairText)
	if match == nil {
		return 0, false
	}
	n, err := strconv.Atoi(match[1])
	if err != nil {
		return 0, false
	}
	return n, true
}

// ParseFlairRange extracts the min-max trade range from a flair template text.
func ParseFlairRange(text string) (min, max int, ok bool) {
	match := flairTemplatePattern.FindStringSubmatch(text)
	if match == nil {
		return 0, 0, false
	}
	min, err1 := strconv.Atoi(match[2])
	max, err2 := strconv.Atoi(match[3])
	if err1 != nil || err2 != nil {
		return 0, 0, false
	}
	return min, max, true
}

// FormatFlairFromTemplate replaces the tracked trade range with the exact count.
func FormatFlairFromTemplate(flairTemplate string, count int) string {
	loc := flairTemplatePattern.FindStringSubmatchIndex(flairTemplate)
	if loc == nil {
		return flairTemplate
	}
	// group 1 is the "min-max" span
	start, end := loc[2], loc[3]
	return flairTemplate[:start] + strconv.Itoa(count) + flairTemplate[end:]
}

// ShouldIncludeComment decides whether a polled comment should be processed.
func ShouldIncludeComment(submissionID, currentSubmissionID string, isRoot bool, bodyLower string) bool {
	if isRoot {
		return submissionID != currentSubmissionID
	}
	return strings.Contains(bodyLower, "confirmed") || strings.Contains(bodyLower, "approved")
}

// IsPossibleWatermarkGap detects when scanning reached the listing limit without finding known IDs.
func IsPossibleWatermarkGap(hadInitialWatermark, foundSeen, listingExhausted bool, scannedCount, gapThreshold int) bool {
	return hadInitialWatermark && !foundSeen && listingExhausted && scannedCount >= gapThreshold
}

// EvaluateConfirmation validates a confirmation comment given pre-fetched context.
func EvaluateConfirmation(comment models.CommentData, ctx models.ConfirmationContext) models.ValidationResult {
	if comment.IsRoot {
		return models.ValidationResult{Valid: false}
	}
	if !ctx.ParentExists || ctx.ParentIsBanned {
		return models.ValidationResult{Valid: false}
	}
	if !ctx.ParentIsProcessable {
		return models.ValidationResult{Valid: false}
	}
	if ctx.ParentAuthorName == comment.AuthorName {
		return models.ValidationResult{Valid: false}
	}

	commentBody := strings.ToLower(comment.Body)

	// Mod approval path: reply to a confirmation (non-root parent).
	if !ctx.ParentIsRoot {
		if strings.Contains(commentBody, "approved") && ctx.IsModerator {
			if ctx.GrandparentExists && ctx.GrandparentIsRoot {
				return models.ValidationResult{
					Valid:            true,
					IsModApproval:    true,
					ParentAuthor:     ctx.GrandparentAuthorName,
					Confirmer:        ctx.ParentAuthorName,
					ParentCommentID:  ctx.GrandparentID,
					ReplyToCommentID: ctx.ParentID,
				}
			}
		}
		return models.ValidationResult{Valid: false}
	}

	if !IsConfirmingTrade(commentBody) {
		return models.ValidationResult{Valid: false}
	}

	if ctx.ParentIsSaved {
		return models.ValidationResult{
			Valid:           false,
			Reason:          "already_confirmed",
			ParentAuthor:    ctx.ParentAuthorName,
			ParentCommentID: ctx.ParentID,
		}
	}

	usernameLower := strings.ToLower(comment.AuthorName)
	if !strings.Contains(ctx.ParentBodyLower, usernameLower) &&
		!strings.Contains(ctx.ParentBodyHTMLLower, usernameLower) {
		return models.ValidationResult{
			Valid:               false,
			Reason:              "cant_confirm_username",
			ParentAuthor:        ctx.ParentAuthorName,
			ParentBodyLower:     ctx.ParentBodyLower,
			ParentBodyHTMLLower: ctx.ParentBodyHTMLLower,
			Confirmer:           usernameLower,
		}
	}

	return models.ValidationResult{
		Valid:            true,
		ParentAuthor:     ctx.ParentAuthorName,
		Confirmer:        comment.AuthorName,
		ParentCommentID:  ctx.ParentID,
		ReplyToCommentID: comment.ID,
	}
}

// FindFlairTemplate finds the flair template matching trade count and mod status.
func FindFlairTemplate(templates []models.FlairTemplate, tradeCount int, isModerator bool) *models.FlairTemplate {
	for i := range templates {
		t := &templates[i]
		if t.Min <= tradeCount && tradeCount <= t.Max && t.ModOnly == isModerator {
			return t
		}
	}
	return nil
}
